pkg/quarto/http.go

```diff
@@ -0,0 +1,72 @@
+package quarto
+
+import (
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/navikt/nada-backend/pkg/database"
+	"github.com/navikt/nada-backend/pkg/gcs"
+)
+
+type Handler struct {
+	repo       *database.Repo
+	gcsClient  gcs.GCS
+	bucketName string
+}
+
+func NewHandler(repo *database.Repo, gcsClient gcs.GCS) *Handler {
+	return &Handler{
+		repo:      repo,
+		gcsClient: gcsClient,
+	}
+}
+
+func (h *Handler) Redirect(w http.ResponseWriter, r *http.Request) {
+	pathParts := strings.Split(r.URL.Path, "/")
+	qID := pathParts[2]
+	r.URL.Path = "/quarto/"
+
+	objPath, err := h.gcsClient.GetIndexHtmlPath(r.Context(), qID)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+
+	http.Redirect(w, r, objPath, http.StatusSeeOther)
+}
+
+func (h *Handler) GetObject(w http.ResponseWriter, r *http.Request) {
+	path := strings.TrimLeft(r.URL.Path, "/quarto")
+
+	objBytes, err := h.gcsClient.GetObject(r.Context(), path)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	switch {
+	case strings.HasSuffix(path, ".html"):
+		w.Header().Add("content-type", "text/html")
+	case strings.HasSuffix(path, ".css"):
+		w.Header().Add("content-type", "text/css")
+	case strings.HasSuffix(path, ".js"):
+		w.Header().Add("content-type", "application/javascript")
+	case strings.HasSuffix(path, ".json"):
+		w.Header().Add("content-type", "application/json")
+	}
+
+	w.Write(objBytes)
+}
+
+func (h *Handler) QuartoMiddleware(next http.Handler) http.Handler {
+	regex, _ := regexp.Compile(`[\n]*\.[\n]*`) // check if object path has file extension
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !regex.MatchString(r.URL.Path) {
```

Comment:
I dont quite understand this code: if the path does not match the pattern of a file name with extension, we will redirect and get the object for the user? It seems we should do so when the path matches?

Reply:
I'm open for suggestions on how to solve this.

What I wanted to achieve was that when accessing e.g. /quarto/<id> we get redirected to the index.html file (or some other .html file should index not exist) for that quarto id.

So when attempting to access a file (e.g. .js, .css, .html etc.) the middleware shall ensure that the GetObject handler is called. If it does not match a file it shall search for the index file using the path prefix and redirect.

Again, I'm open for better alternatives on to how solve this, but I'm not quite sure what you mean when you're saying we should redirect when the path matches?